tests/CallbacksTest.test.nut

```diff
@@ -0,0 +1,160 @@
+      this.cm.connect();
+    }.bindenv(this));
+    this.assertTrue(server.isconnected(), "should be connected!");
+    this.cm.disconnect();
+    this.assertTrue(!server.isconnected(), "should NOT be connected!");
+    return Promise(function (resolve, reject) {
+      imp.wakeup(this.TIMEOUT_2, function () {
+        try {
+          this.info("should be online now");
+          this.assertTrue(server.isconnected(), "should be connected again!");
+          resolve();
+        } catch (e) {
+          this.cm.connect(); 
+          reject(e);
+        }
+      }.bindenv(this));
+    }.bindenv(this));
+  }
+
+ /*
+  * sets onConnect callback and disconnects device using CM
+  *
+  */
+  function testOnConnectAsync() {
+    local flag = 0;
+    cm.onConnect(function() {
+      flag++;
+    }.bindenv(this));
+    this.assertTrue((flag == 0), "should NOT be true yet");
+    this.assertTrue(server.isconnected(), "should be connected!");
+    this.cm.disconnect();
+    this.assertTrue(!server.isconnected(), "should NOT be connected!");
+    return Promise(function (resolve, reject) {
+      imp.wakeup(0, function() {
+        this.cm.connect();
+      }.bindenv(this));
+      imp.wakeup(this.TIMEOUT_2, function () {
+          try {
+            this.info("should be online now");
+            this.assertTrue(server.isconnected(), "should be connected again!");
+            this.assertTrue((flag == 1), "should be true now");
+            resolve();
+          } catch (e) {
+            reject(e);
+          }
+      }.bindenv(this));
+    }.bindenv(this));
+  }
+
+ /*
+  * sets onNextConnect callback and disconnects device using CM
+  *
+  */
+  function testOnNextConnectAsync() {
+    local flag = 0;
+    cm.onNextConnect(function() {
+      flag++;
+    }.bindenv(this));
+    this.assertTrue((flag == 0), "should NOT be true yet");
+    this.assertTrue(server.isconnected(), "should be connected!");
+    this.cm.disconnect();
+    this.assertTrue(!server.isconnected(), "should NOT be connected!");
+    return Promise(function (resolve, reject) {
+      imp.wakeup(this.TIMEOUT_1, function() {
+        this.cm.connect();
+      }.bindenv(this));
+      imp.wakeup(this.TIMEOUT_2, function () {
+        try {
+          this.assertTrue(server.isconnected(), "should be connected again!");
+          this.assertTrue((flag == 1), "should be true now");
+          resolve();
+        } catch (e) {
+          reject(e);
+        }
+      }.bindenv(this));
+    }.bindenv(this));
+  }
+
+ /*
+  * removes onConnect callback checks it is not firing any more
+  *
+  */
+  function testOnConnectRemovalAsync() {
+    local flag = 0;
+    cm.onConnect(function() {
+      flag++;
+    }.bindenv(this));
+    //cm.connect returns false iff it's already in process of connection
+    this.assertTrue(cm.connect(), "should NOT be connecting now");
+    return Promise(function (resolve, reject) {
+      imp.wakeup(this.TIMEOUT_1, function() {
```

Comment:
The problem with this hardcoded times is that it may take more than 20 seconds sometime for a imp to connect depending on the network state and RSSI. And these tests will keep intermittently failing - you'll need to restart them all the time. More reliably (but still depends on the network of course) would be to have a global timeout per test (big enough to make sure it all passes under normal conditions) and instead of 
```
imp.wakeup(this.TIMEOUT_1, function() {...})
```
do Promise chaining.
Another option, which we discussed, but is missing here is to supply integration tests with unit-tests.